### DISCUSSION TITLE
How to drop or past local image in REACT

### DISCUSSION THREAD

**Discussion by neostom432** [ueberdosis/tiptap#2977] [General Questions] [answered]
Hello, I'm trying to make my user can attatch local image to editor by drag-and-drop(or paste) local file but I don't know how to handle it.

Should I add event handler to `EditorContent`?

If so, what kind of event handler should I use?
I've tried `onPaste` and `onDrop` but it seems not working.

Could you please give any clue?

**Comment by sereneinserenade** [comment]
You can convert it to a base64 string and add it to the `src` attribute. let me know if you want an example. 

https://prosemirror.net/examples/upload/ is the demo, and code is here https://github.com/ProseMirror/website/blob/f9e76c85f09132876c904223798cf81464c3e60c/example/upload/index.js take a look at this.

**Reply by sereneinserenade** [reply]
@neostom432 I've added more details, please have a look!

**Reply by neostom432** [reply]
Thank you!
Can I ask your example support drag-and-drop and paste both?

**Reply by sereneinserenade** [reply]
this is the precise bit you're looking for

https://github.com/ProseMirror/website/blob/master/example/upload/index.js#L40-L44

**Reply by sereneinserenade** [reply]
I think paste you can just do with a prosemirror plugin https://prosemirror.net/docs/ref/#state.Plugin_System still let me know if you want me to provide an example, I'll do that.

**Reply by neostom432** [reply]
Thanks for your demo but point of my question is how to handle paste/and drag-and-drop to editor.
Your demo looks like how to handle local image which I know how.

**Reply by neostom432** [reply]
I'm just start using tiptap so I don't know much about promisemirror.
It seems that i should know promisemirror to handle paste/drag-and-drop.

Is it right?

**Reply by sereneinserenade** [reply]
if you want to handle DOM Events, afaik ProseMirror is the way to go, for my comment extension https://github.com/sereneinserenade/tiptap-comment-extension, I use a prosemirror plugin to handle `click` event. https://github.com/sereneinserenade/tiptap-comment-extension/blob/main/src/components/extension/comment.ts#L69-L76 the same way, you can handle the `drop` event on editor, and since you already know how to handle a `local image`, it shouldn't be an issue I believe.

**Reply by neostom432** [reply]
If I want to handle DOM event for node, your demo is perfect example for me. 

But my case, I think paste/drop event should driven by editor. So I tried add onPaste / onDrop event handler to editor and found out that I tested wrong way before asking this question.


I tested again like this
`
      <EditorContent
        editor={editor}
        className={cx('editor-container')}
        onPaste={(e) => {
          console.log('paste');
          console.log(e);
        }}
        onDrop={(e) => {
          e.preventDefault();
          e.stopPropagation();
          console.log('DROP2');
          console.log(e);
        }}
      />
`

When copy image to clipboard, paste event is triggered.
And drag-and-drop image from local to editor, drop event is triggered and download(default action) is prevented.

Thanks to give me advice.

( Regardless of the question, your comment extension is super great! I hope I can make cool extension like you :) )

**Reply by neostom432** [reply]
@sereneinserenade 
Could you paste your last answer and make it new comment? I want to pick your last comment as answer.

**Reply by sereneinserenade** [reply]
You can register the editor-level events while initiating the editor. take a look here https://tiptap.dev/api/editor#editor-props

**Reply by neostom432** [reply]
I have last question...
I'm really new in Tiptap and ProseMirror and I usually use Slate which is fully React module.

So it is so confuse... In react, there are specific flow like component mounted, component props updated, so I can handle node using that.
But here, i cannot fully understand when and where should I do. 

I mean, for me, two things make me think that I should add event to editor.
1. paste and drop events are emitted through editor
2. Image is not rendered yet 

But extensions looks like add event handler to itself like what you share to me.
Are there event called after node added?

Say, I'm adding drop event to my custom image extension which extend image extention. When drag-and-drop local image file to editor, is this event called?

I tried add onPaste event to editor and add image in that callback but image is shown and change to empty paragraph immediately..

Here is my code
```
export default function TipTap() {
  const editor = useEditor({
    extensions: [
      StarterKit,
      Image,
      Mention.configure({
        HTMLAttributes: {
          class: cx('mention'),
        },
        renderLabel({ options, node }) {
          return `${options.suggestion.char}${
            node.attrs.label ?? node.attrs.id
          }`;
        },
        suggestion: MentionSuggestion,
      }),
    ],
    content: '',
    editorProps: {
      attributes: {
        class: cx('editor'),
      },
    },
    onUpdate: ({ editor: currentEditor }) => {
      const html = currentEditor.getHTML();
    },
  });

  const addImage = (data: DataTransfer) => {
    const text = data.getData('text/plain');
    const { files } = data;
    console.log(files);

    if (files && files.length > 0) {
      for (const file of Array.from(files)) {
        const [mime] = file.type.split('/');

        if (mime === 'image') {
          const url = URL.createObjectURL(file);
          console.log('IMAGE URL  ' + url);
          editor?.chain().focus().setImage({ src: url }).run();
        }
      }
    }
  };

  useEffect(() => {}, [editor]);

  return (
    <div className={cx('container')}>
      <MenuBar editor={editor} />
      <EditorContent
        editor={editor}
        className={cx('editor-container')}
        onPaste={(e) => {
          console.log('paste');
          addImage(e.clipboardData);
        }}
        onDrop={(e) => {
          e.preventDefault();
          e.stopPropagation();
          console.log('DROP2');
          addImage(e.dataTransfer);
        }}
      />
    </div>
  );
}
```

**Comment by sereneinserenade** [comment, marked as answer]
if you want to handle DOM Events, afaik ProseMirror is the way to go, for my comment extension https://github.com/sereneinserenade/tiptap-comment-extension, I use a prosemirror plugin to handle click event. https://github.com/sereneinserenade/tiptap-comment-extension/blob/main/src/components/extension/comment.ts#L69-L76 the same way, you can handle the drop event on editor, and since you already know how to handle a local image, it shouldn't be an issue I believe.

@neostom432 here you go 😇😁
